docs/testing/testing.md

```diff
@@ -0,0 +1,293 @@
+
+We can keep writing tests, and improving our function but maybe you have noticed that the first two tests are exactly the same, and last two too. Instead of 
+writing them separately we can use some tools that pytest provide and our tests
+will look cleaner. 
+
+For the first case, where we want the result of the dot product to match an
+expected value we will re-designed our test in the following way:
+
+```python
+@pytest.mark.parametrize("a,b,expected",[
+    ([0,0],[0,0], 0), 
+    ([0,1],[1,0],0),
+    ([1,2],[2,-1],0),])
+
+def test_assert_equal(a,b,expected):
+    assert dot(a,b)==expected
+```
+
+In this case we have 3 cases `(a,b,expected)` in a list that will go over our 
+test. After we run the tests in the terminal we'll see:  
+
+```text
+============================= test session starts ==============================
+platform linux -- Python 3.5.2, pytest-3.0.5, py-1.4.31, pluggy-0.4.0
+rootdir: /home/naty/tmp/testing, inifile: 
+collected 5 items 
+
+dot_product.py .....
+
+=========================== 5 passed in 0.02 seconds ===========================
+```
+Following the same idea we can compact the two last tests by doing this:
+
+```python
+@pytest.mark.parametrize("a,b", [
+    ([1,2],[1,2,3]),
+    ([1,2,3],[42,7]),])
+
+def test_for_assert_error(a,b):
+    with pytest.raises(AssertionError):
+        dot(a,b)
+```
+
+Again you should get 5 tests passed, let's see:
+
+```text
+============================= test session starts ==============================
+platform linux -- Python 3.5.2, pytest-3.0.5, py-1.4.31, pluggy-0.4.0
+rootdir: /home/naty/tmp/testing, inifile: 
+collected 5 items 
+
+dot_product.py .....
+
+=========================== 5 passed in 0.03 seconds ===========================
+```
+
+## Keeping the tests in a separate file
+
+For a better organization of your code we strongly recomment having your tests
+in a separate file. In the example we've been working a way to do that can be:
+
+In a file called `dot_product.py` you save just the function:
+
+```python
+def dot (a,b):
+    assert len(a)==len(b)  #This is the line we added
+    res = 0
+    for i in range(len(a)):
+        res += a[i] * b[i]
+    return res
+```
+
+In a separate file that can be named `test_dot.py` you put the tests, and the imports that you need, in this
+case you will `import pytest` and `from dot_product import dot`, therefore your file will look likes:
+
+```python
+import pytest
+from dot_product import dot
+
+@pytest.mark.parametrize("a,b,expected",[
+    ([0,0],[0,0], 0), 
+    ([0,1],[1,0],0),
+    ([1,2],[2,-1],0),])
+
+def test_assert_equal(a,b,expected):
+    assert dot(a,b)==expected
+
+@pytest.mark.parametrize("a,b", [
+    ([1,2],[1,2,3]),
+    ([1,2,3],[42,7]),])
+
+def test_for_assert_error(a,b):
+    with pytest.raises(AssertionError):
+        dot(a,b)
+``` 
+
+Both files should be in the same location and then you just run in you terminal
+`pytest test_dot.py`.
```

Comment:
If there's a test file that begins with `test_` then you only need to run `pytest` without any other arguments and pytest will detect it.